core/trino-main/src/main/java/io/trino/execution/QueryPreparer.java

```diff
@@ -65,19 +68,39 @@ public PreparedQuery prepareQuery(Session session, Statement wrappedStatement)
             prepareSql = Optional.of(session.getPreparedStatementFromExecute(executeStatement));
             statement = sqlParser.createStatement(prepareSql.get(), createParsingOptions(session));
         }
-
-        if (statement instanceof ExplainAnalyze explainAnalyzeStatement) {
+        else if (statement instanceof ExecuteImmediate executeImmediateStatement) {
+            // TODO: consider adding dedicated logic to deal with ExecuteImmediate in the analyzer
+            try {
+                statement = sqlParser.createStatement(executeImmediateStatement.getStatement(), createParsingOptions(session));
+            }
+            catch (ParsingException e) {
+                int embeddedStatementLine = executeImmediateStatement.getStatementLocation().getLineNumber();
+                int embeddedStatementColumn = executeImmediateStatement.getStatementLocation().getColumnNumber();
+                ParsingException exceptionWithUpdatedLocation = new ParsingException(
```

Comment:
Also, this won't work for a test such as this, as it only handles parsing errors:

```
        assertTrinoExceptionThrownBy(() -> assertions.query("EXECUTE IMMEDIATE 'SELECT * FROM tiny.tpch.orders'"))
                .hasMessageMatching("line 1:34: Catalog 'tiny' does not exist");
```

Reply:
If the SqlParser could take a "relative start position", it would solve both issues:
- parsing exceptions would have the correct location,
- the created AST would have the correct locations, so the Analyzer errors would be positioned correctly, too.

Reply:
>Also, this won't work for a test such as this, as it only handles parsing errors:

BTW @martint we have the same problem with `EXECUTE`:
```
trino> prepare test from select * from foo;
PREPARE
trino> execute test;
Query 20230510_184850_00010_fczhd failed: line 3:3: Schema must be specified when session schema is not set
```
It's even worse, as the line/column in error message refer to how the statement was formatter by the SqlFormatter, not to how it was originally written by the user.

Probably in this case we should skip the position altogether, or place it at the beginning of the `EXECUTE` statement, or maybe include the executed statement in the message to gie context for the error.